### PR TITLE
[FLINK-10599][Documentation] Provide documentation for the modern kafka connector

### PR DESCRIPTION
**Pull request by yanghua**
## What is the purpose of the change

*This pull request provides documentation for the modern kafka connector*

## Brief change log

  - *Provide documentation for the modern kafka connector*

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / **docs** / JavaDocs / not documented)

### CODE REVIEW

**Comment by yanghua** [comment]
Hi @aljoscha I provided a simple introduction and documentation for the modern kafka connector. Considering that I am not a native English user, you are welcome to point out where the documentation is inappropriate. Also, if it is missing something, pelase feel free to tell me. Thank you.

File: docs/dev/connectors/kafka.md [hunk 1]
```diff
@@ -100,6 +108,31 @@ Note that the streaming connectors are currently not part of the binary distribu
...
 * If the Kafka and Zookeeper servers are running on a remote machine, then the `advertised.host.name` setting in the `config/server.properties` file must be set to the machine's IP address.
 
+## Modern Kafka Connector
+
+Starting with Flink 1.7, it will provide a new connector for the modern Kafka. It is different from any previous Kafka connector, it does not bind a fixed version of Kafka client, but it will track and support the latest version of Kafka Client.
+The modern Kafka connector does not rely on any older version of the kafka connector and relies only on the underlying implementation of the kafka connector, which allows it to get rid of the historical burden and keep up with the latest version of the Kafka client. 
+
+Note that if your version of Kafka Server exceeds 0.11, then we recommend you use the modern Kafka connector. The specific version of the kafka connector will continue to maintain and fix related issues. When you use the old version of Kafka (for example, 0.11, 0.10, 0.9, 0.8), we still recommend that you use the connector corresponding to the corresponding version, because they are precisely matched.
```
Comment (eliaslevy) [inline docs/dev/connectors/kafka.md]:
```suggestion
If your Kafka broker version is 1.0.0 or newer, you should use this Kafka connector. If you use an older version of Kafka (0.11, 0.10, 0.9, or 0.8), you should use the connector corresponding to the broker version.
```

File: docs/dev/connectors/kafka.md [hunk 2]
```diff
@@ -100,6 +108,31 @@ Note that the streaming connectors are currently not part of the binary distribu
...
+
+### Compatibility
+
+The compatibility of the modern Kafka connector is completely dependent on the backward compatibility guarantee of the Kafka Client API (Kafka supports the latest Client API to access the old version of Kafka Server. For more details on Kafka compatibility, please refer to the [official Kafka documentation](https://kafka.apache.org/protocol.html#protocol_compatibility)). Therefore, as long as the connector is upgraded and supported to the latest Kafka Client version, it can provide the same compatibility guarantee as the Kafka Client.
```
Comment (eliaslevy) [inline docs/dev/connectors/kafka.md]:
```suggestion
The the modern Kafka connector is compatible with older and newer Kafka brokers through the compatibility guarantees of the Kafka client API and broker. The modern Kafka client is compatible with broker versions 0.10.0 or later, depending on the features used. Fo details on Kafka compatibility, please refer to the [Kafka documentation](https://kafka.apache.org/protocol.html#protocol_compatibility).
```

File: docs/dev/connectors/kafka.md [hunk 3]
```diff
@@ -100,6 +108,31 @@ Note that the streaming connectors are currently not part of the binary distribu
...
+
+### Usage
+
+The use of the modern Kafka connector is very simple. We only need to introduce dependencies on it:
```
Comment (eliaslevy) [inline docs/dev/connectors/kafka.md]:
```suggestion
The use of the modern Kafka connector add a dependency to it:
```

File: docs/dev/connectors/kafka.md [hunk 4]
```diff
@@ -100,6 +108,31 @@ Note that the streaming connectors are currently not part of the binary distribu
...
+
+{% highlight xml %}
+<dependency>
+  <groupId>org.apache.flink</groupId>
+  <artifactId>flink-connector-kafka{{ site.scala_version_suffix }}</artifactId>
+  <version>{{site.version }}</version>
+</dependency>
+{% endhighlight %}
+
+Then instantiate the new source (`FlinkKafkaConsumer`) and sink (`FlinkKafkaProducer`). As for the API, it is fully compatible with other Kafka connectors that already exist. So in theory, you don't need to make any changes to it to make your program work.
```
Comment (eliaslevy) [inline docs/dev/connectors/kafka.md]:
```suggestion
Then instantiate the new source (`FlinkKafkaConsumer`) and sink (`FlinkKafkaProducer`). The API is the backwards compatible with the older Kafka connectors.
```

File: docs/dev/connectors/kafka.md [hunk 5]
```diff
@@ -80,6 +80,14 @@ For most users, the `FlinkKafkaConsumer08` (part of `flink-connector-kafka`) is
         <td>0.11.x</td>
         <td>Since 0.11.x Kafka does not support scala 2.10. This connector supports <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-98+-+Exactly+Once+Delivery+and+Transactional+Messaging">Kafka transactional messaging</a> to provide exactly once semantic for the producer.</td>
     </tr>
+    <tr>
+        <td>flink-connector-kafka_2.11</td>
+        <td>1.7.0</td>
+        <td>FlinkKafkaConsumer<br>
+        FlinkKafkaProducer</td>
+        <td>1.x,2.x</td>
```
Comment (eliaslevy) [inline docs/dev/connectors/kafka.md]:
```suggestion
        <td>>= 1.x</td>
```

File: docs/dev/connectors/kafka.md [hunk 6]
```diff
@@ -100,6 +108,31 @@ Note that the streaming connectors are currently not part of the binary distribu
 * Follow the instructions from [Kafka's quickstart](https://kafka.apache.org/documentation.html#quickstart) to download the code and launch a server (launching a Zookeeper and a Kafka server is required every time before starting the application).
 * If the Kafka and Zookeeper servers are running on a remote machine, then the `advertised.host.name` setting in the `config/server.properties` file must be set to the machine's IP address.
 
+## Modern Kafka Connector
```
Comment (eliaslevy) [inline docs/dev/connectors/kafka.md]: I am wondering if instead of "Modern" we should just call it "unversioned".

Comment (yanghua) [reply docs/dev/connectors/kafka.md]: I don't object, what do you think? cc @aljoscha

**Comment by yanghua** [comment]
Hi @eliaslevy Thank you for reviewing it, I have adopted most of your suggestions.

File: docs/dev/connectors/kafka.md [hunk 7]
```diff
@@ -100,6 +108,30 @@ Note that the streaming connectors are currently not part of the binary distribu
...
+
+Starting with Flink 1.7, there is a new Kafka connector that does not track a specific Kafka major version. Rather, it tracks the latest version of Kafka at the time of the Flink release.
+
+If your Kafka broker version is 1.0.0 or newer, you should use this Kafka connector. If you use an older version of Kafka (0.11, 0.10, 0.9, or 0.8), you should use the connector corresponding to the broker version.
+
+### Compatibility
+
+The the modern Kafka connector is compatible with older and newer Kafka brokers through the compatibility guarantees of the Kafka client API and broker. The modern Kafka client is compatible with broker versions 0.10.0 or later, depending on the features used. Fo details on Kafka compatibility, please refer to the [Kafka documentation](https://kafka.apache.org/protocol.html#protocol_compatibility).
```
Comment (zentol) [inline docs/dev/connectors/kafka.md]: double "the"; typo: Fo -> For

File: docs/dev/connectors/kafka.md [hunk 8]
```diff
@@ -100,6 +108,30 @@ Note that the streaming connectors are currently not part of the binary distribu
...
+
+### Usage
+
+The use of the modern Kafka connector add a dependency to it:
+
+{% highlight xml %}
+<dependency>
+  <groupId>org.apache.flink</groupId>
+  <artifactId>flink-connector-kafka{{ site.scala_version_suffix }}</artifactId>
+  <version>{{site.version }}</version>
```
Comment (zentol) [inline docs/dev/connectors/kafka.md]: sync number of spaces around `site.version`

**Comment by yanghua** [comment]
@zentol updated.

**Comment by yanghua** [comment]
@zentol @aljoscha @pnowojski Can you give a final check about this documentation?

File: docs/dev/connectors/kafka.md (hunk 6 above)
Comment (pnowojski) [reply docs/dev/connectors/kafka.md]: Both of those names can be confusing for the user. Maybe we should name this section as `Kafka 0.11+ Connector`?

File: docs/dev/connectors/kafka.md [hunk 9]
```diff
@@ -100,6 +108,30 @@ Note that the streaming connectors are currently not part of the binary distribu
...
 
+## Modern Kafka Connector
+
+Starting with Flink 1.7, there is a new Kafka connector that does not track a specific Kafka major version. Rather, it tracks the latest version of Kafka at the time of the Flink release.
+
+If your Kafka broker version is 1.0.0 or newer, you should use this Kafka connector. If you use an older version of Kafka (0.11, 0.10, 0.9, or 0.8), you should use the connector corresponding to the broker version.
+
+### Compatibility
+
+The modern Kafka connector is compatible with older and newer Kafka brokers through the compatibility guarantees of the Kafka client API and broker. The modern Kafka client is compatible with broker versions 0.10.0 or later, depending on the features used. For details on Kafka compatibility, please refer to the [Kafka documentation](https://kafka.apache.org/protocol.html#protocol_compatibility).
```
Comment (pnowojski) [inline docs/dev/connectors/kafka.md]: Have we tested that this new connector works with Kafka 0.10? Definitely not in `exacty_once` mode. Maybe it does work with `at_least_once` mode, but if we haven't tested that, I would refrain from saying it. So I would rephrase this to:

```
The modern Kafka connector is compatible with broker versions 0.11.0 or later
```

File: docs/dev/connectors/kafka.md [hunk 10]
```diff
@@ -80,6 +80,14 @@ For most users, the `FlinkKafkaConsumer08` (part of `flink-connector-kafka`) is
         <td>0.11.x</td>
         <td>Since 0.11.x Kafka does not support scala 2.10. This connector supports <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-98+-+Exactly+Once+Delivery+and+Transactional+Messaging">Kafka transactional messaging</a> to provide exactly once semantic for the producer.</td>
     </tr>
+    <tr>
+        <td>flink-connector-kafka_2.11</td>
+        <td>1.7.0</td>
+        <td>FlinkKafkaConsumer<br>
+        FlinkKafkaProducer</td>
+        <td>>= 1.x</td>
```
Comment (pnowojski) [inline docs/dev/connectors/kafka.md]: `>= 1.0.0`

File: docs/dev/connectors/kafka.md [hunk 11]
```diff
@@ -100,6 +108,30 @@ Note that the streaming connectors are currently not part of the binary distribu
...
+  <artifactId>flink-connector-kafka{{ site.scala_version_suffix }}</artifactId>
+  <version>{{ site.version }}</version>
+</dependency>
+{% endhighlight %}
+
+Then instantiate the new source (`FlinkKafkaConsumer`) and sink (`FlinkKafkaProducer`). The API is the backwards compatible with the older Kafka connectors.
+
 ## Kafka Consumer
 
 Flink's Kafka consumer is called `FlinkKafkaConsumer08` (or `09` for Kafka 0.9.0.x versions, etc.). It provides access to one or more Kafka topics.
```
Comment (pnowojski) [inline docs/dev/connectors/kafka.md]: Should this line also be adjusted? I think we should also mention here that "modern" connector brakes the naming convention.

Same applies to similar sentence later for the producer:
```
Flink’s Kafka Producer is called FlinkKafkaProducer011 (or 010 for Kafka 0.10.0.x versions, etc.). It allows writing a stream of records to one or more Kafka topics.
```

Also in `Kafka Producers and Fault Tolerance` section, rename the title `Kafka 0.11` to `Kafka 0.11 and newer`, and also adopt the content that section (the name `FlinkKafkaProducer011 ` is used there frequently and with modern connector it should also be adjusted).

**Comment by yanghua** [comment]
@pnowojski thanks for your suggestion, updated.

**Comment by yanghua** [comment]
@pnowojski updated.
